File: Applications/Python/LinkedList/CopyListWithRandomPointer.py
# 138. Copy List with Random Pointer


class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

class Solution:
    def copyRandomList(self, head):
        copies = {None: None}

        curr = head
        while curr:
            cp = Node(curr.val)
            copies[curr] = cp
            curr = curr.next

        curr = head
        while curr:
            cp = copies[curr]
            cp.next = copies[curr.next]
            cp.random = copies[curr.random]
            curr = curr.next

        return copies[head]

File: Applications/CSharp/LinkedList/CopyListWithRandomPointer.cs
// 138. Copy List with Random Pointer

class Node
{
    public int val;
    public Node next;
    public Node random;

    public Node(int _val)
    {
        val = _val;
        next = null;
        random = null;
    }
}

class Solution
{
    public Node CopyRandomList(Node head)
    {
        if (head == null)
            return null;

        var map = new Dictionary<int, Node>();

        var curr = head;
        while (curr != null)
        {
            var cp = new Node(curr.val);
            map.Add(curr.GetHashCode(), cp);
            curr = curr.next;
        }

        curr = head;
        while (curr != null)
        {
            var cp = map[curr.GetHashCode()];
            if (curr.next != null)
                cp.next = map[curr.next.GetHashCode()];
            if (curr.random != null)
                cp.random = map[curr.random.GetHashCode()];
            curr = curr.next;
        }

        return map[head.GetHashCode()];
    }
}